hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzonePrefixPathImpl.java

```diff
@@ -0,0 +1,153 @@
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+import static org.apache.hadoop.ozone.om.exceptions.OMException.ResultCodes.KEY_NOT_FOUND;
+
+public class OzonePrefixPathImpl implements OzonePrefixPath {
+
+  private String volumeName;
+  private String bucketName;
+  private KeyManager keyManager;
+  // TODO: based on need can make batchSize configurable.
+  private int batchSize = 1000;
+  private OzoneFileStatus pathStatus;
+
+  public OzonePrefixPathImpl(String volumeName, String bucketName,
+      String keyPrefix, KeyManager keyManagerImpl) throws IOException {
+    this.volumeName = volumeName;
+    this.bucketName = bucketName;
+    this.keyManager = keyManagerImpl;
+
+    OmKeyArgs omKeyArgs = new OmKeyArgs.Builder()
+        .setVolumeName(volumeName)
+        .setBucketName(bucketName)
+        .setKeyName(keyPrefix)
+        .setRefreshPipeline(false)
+        .build();
+    try {
+      pathStatus = keyManager.getFileStatus(omKeyArgs);
+    } catch (OMException ome) {
+      // In existing code non-FSO code, ozone client delete and rename expects
+      // KNF error code. So converting FNF to KEY_NOT_FOUND error code.
+      if (ome.getResult() == OMException.ResultCodes.FILE_NOT_FOUND) {
+        throw new OMException(ome.getMessage(), KEY_NOT_FOUND);
+      }
+      throw ome;
+    }
+  }
+
+  @Override
+  public OzoneFileStatus getOzoneFileStatus() {
+    return pathStatus;
+  }
+
+  @Override
+  public Iterator<? extends OzoneFileStatus> getChildren(String keyPrefix)
+      throws IOException {
+
+    return new PathIterator(keyPrefix);
+  }
+
+  class PathIterator implements Iterator<OzoneFileStatus> {
+    private Iterator<OzoneFileStatus> currentIterator;
+    private String keyPrefix;
+    private OzoneFileStatus currentValue;
+
+    /**
+     * Creates an Iterator to iterate over all sub paths of the given keyPrefix.
+     *
+     * @param keyPrefix
+     */
+    PathIterator(String keyPrefix) throws IOException {
+      this.keyPrefix = keyPrefix;
+      this.currentValue = null;
+      List<OzoneFileStatus> statuses = getNextListOfKeys("");
+      if (statuses.size() == 1) {
+        OzoneFileStatus keyStatus = statuses.get(0);
+        if (keyStatus.isFile() && StringUtils.equals(keyPrefix,
+            keyStatus.getTrimmedName())) {
+          throw new OMException("Invalid keyPrefix: " + keyPrefix +
+              ", file type is not allowed, expected directory type.",
+              OMException.ResultCodes.INVALID_KEY_NAME);
+        }
+      }
+      this.currentIterator = statuses.iterator();
+    }
+
+    @Override
+    public boolean hasNext() {
+      if (!currentIterator.hasNext() && currentValue != null) {
+        try {
+          currentIterator =
+              getNextListOfKeys(currentValue.getTrimmedName()).iterator();
+        } catch (IOException e) {
+          throw new RuntimeException(e);
```

Comment:
Can we avoid throw RuntimeException here?

Reply:
OK. Then I will return false if any exception thrown by `getNextListOfKeys(currentValue.getTrimmedName())`  function call.